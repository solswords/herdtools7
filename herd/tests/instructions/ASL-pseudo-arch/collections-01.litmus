ASL collections-01

{
  x = 0;
  y = 0;
  0: X1= x;
  0: X2= y;
}

type MyCollection of collection {
  field1: bits(8),
  field2: bits(8),
};

constant one : bits(64) = 1[63:0];
var global_collection: MyCollection;

func T0(x:bits(64), y: bits(64))
begin
  write_memory{64}(x, one);
  write_memory{64}(y, one);
end;

func T1(x:bits(64), y: bits(64))
begin
  global_collection.field2 = Zeros{64};
  let a = UInt(read_memory{64}(y));
  global_collection.field1 = a[8:0]; // WRITE field1
  let b = UInt(read_memory{64}(x + UInt(global_collection.field2))); // READ field2
end;

func main() => integer
begin
  let x = read_register(1);
  let y = read_register(2);

  T0(x, y);
  T1(x, y);

  return 0;
end;

exists (0: T1.0.a = 1 /\ 0: T1.0.b = 0)

