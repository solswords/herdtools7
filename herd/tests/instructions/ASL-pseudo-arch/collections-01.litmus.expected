Test collections-01 Allowed
States 4
0:T1.0.a=0; 0:T1.0.b=0;
0:T1.0.a=0; 0:T1.0.b=1;
0:T1.0.a=1; 0:T1.0.b=0;
0:T1.0.a=1; 0:T1.0.b=1;
Ok
Witnesses
Positive: 1 Negative: 3
Condition exists (0:T1.0.a=1 /\ 0:T1.0.b=0)
Observation collections-01 Sometimes 1 3
Hash=4406602ca1742f126d5fc2909064a894

